src/video_core/renderer_vulkan/vk_pipeline_cache.h

```diff
@@ -33,26 +42,71 @@ class PipelineCache {
 
 private:
     void RefreshGraphicsKey();
-    void DumpShader(std::span<const u32> code, u64 hash, Shader::Stage stage, std::string_view ext);
+    void RefreshComputeKey();
+    void DumpShader(std::span<const u32> code, u64 hash, Shader::Stage stage, size_t perm_idx,
+                    std::string_view ext);
+
+    vk::ShaderModule CompileModule(Shader::Info& info, std::span<const u32> code, size_t perm_idx,
+                                   u32& binding);
+
+    std::tuple<const Shader::Info*, vk::ShaderModule, u64> GetProgram(const auto* pgm,
+                                                                      Shader::Stage stage,
+                                                                      u32& binding) {
+        // Fetch program for binaryinfo hash.
+        const auto* bininfo = Liverpool::GetBinaryInfo(*pgm);
+        const u64 hash = bininfo->shader_hash;
+        auto [it_pgm, new_program] = program_cache.try_emplace(hash);
+        u64 stage_key{};
+        if (new_program) {
+            // Create a new program and a module with current runtime state.
+            const VAddr pgm_base = pgm->template Address<VAddr>();
+            auto program = program_pool.Create();
+            program->info = MakeShaderInfo(stage, pgm->user_data, pgm_base, hash, liverpool->regs);
+            u32 start_binding = binding;
+            const auto module = CompileModule(program->info, pgm->Code(), 0, start_binding);
+            stage_key = program->info.GetStageSpecializedKey(binding);
+            program->modules.emplace_back(stage_key, module);
+            it_pgm.value() = program;
+        } else {
+            stage_key = it_pgm->second->info.GetStageSpecializedKey(binding);
+        }
+
+        Program* program = it_pgm->second;
+        const auto& info = program->info;
+
+        // Compile specialized module with current runtime state.
+        const auto it = std::ranges::find(program->modules, stage_key, &Program::Module::first);
+        if (it == program->modules.end()) {
+            auto new_info = MakeShaderInfo(stage, pgm->user_data, info.pgm_base, info.pgm_hash,
+                                           liverpool->regs);
+            const size_t perm_idx = program->modules.size();
+            const auto module = CompileModule(new_info, pgm->Code(), perm_idx, binding);
+            program->modules.emplace_back(stage_key, module);
```

Comment:
I can see how program->modules.end() points to the freshly emplaced element so you just reuse the iterator when returning from this function.

Some shaders cause the small_vector to allocate on the heap, invalidating this iterator and propagating VK_NULL_HANDLE to the pipeline creation functions.